File: backend/core/api/app/services/workflow_template_service.py
# backend/core/api/app/services/workflow_template_service.py
#
# Persistent storage and import validation for opaque, client-encrypted Workflow
# template projections. Runtime workflow data remains authoritative and Vault
# encrypted; this service never decrypts projection ciphertext or template keys.
#
# Spec: docs/specs/workflows-v1/spec.yml (TASK-5, T-PYTEST-008)

from __future__ import annotations

import time
import uuid
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, ValidationError

from backend.core.api.app.services.workflow_models import WorkflowDetail, WorkflowGraph, WorkflowNodeType
from backend.core.api.app.services.workflow_service import WorkflowService


FORBIDDEN_TEMPLATE_FIELD_PARTS = {
    "token",
    "refreshtoken",
    "accesstoken",
    "authtoken",
    "bearertoken",
    "secret",
    "credential",
    "accountid",
    "connectionid",
    "connectedaccountid",
    "provideruserid",
    "webhooksecret",
    "apikey",
    "password",
    "vault",
    "grant",
    "runid",
    "versionid",
    "workflowid",
    "nextrunat",
    "claim",
    "wait",
    "output",
    "providerresponse",
    "sourcechatid",
    "encryptedgraphblobref",
    "encryptedcontentref",
    "fragmentkey",
    "shortkey",
    "templatekey",
}


class WorkflowTemplateProjectionError(ValueError):
    """Raised when an opaque template projection cannot be persisted."""


class WorkflowTemplateProjectionStaleError(WorkflowTemplateProjectionError):
    """Raised when a client snapshot predates the current runtime workflow."""


class WorkflowTemplateImportError(ValueError):
    """Raised when a client-decrypted template is not safe to import."""


class WorkflowTemplateProjectionRecord(BaseModel):
    """Opaque persistent record used for owner cross-device and share flows."""

    model_config = ConfigDict(extra="forbid")

    template_id: str = Field(min_length=1, max_length=200)
    workflow_id: str = Field(min_length=1, max_length=200)
    owner_hash: str = Field(min_length=1)
    source_version: int = Field(ge=1)
    projection_schema_version: int = Field(ge=1)
    ciphertext: str = Field(min_length=1, max_length=100_000)
    ciphertext_checksum: str = Field(min_length=1, max_length=200)
    owner_wrapped_key: str = Field(min_length=1, max_length=100_000)
    created_at: int
    updated_at: int
    revoked_at: int | None = None


class WorkflowTemplateImportPayload(BaseModel):
    """Portable plaintext accepted only after client-side decryption."""

    model_config = ConfigDict(extra="forbid")

    template_version: int = Field(ge=1)
    title: str = Field(min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=2_000)
    trigger_template: dict[str, Any]
    node_templates: list[dict[str, Any]] = Field(default_factory=list)
    edge_templates: list[dict[str, Any]] = Field(default_factory=list)
    variables_schema: dict[str, Any] = Field(default_factory=dict)
    required_capabilities: list[str] = Field(default_factory=list)
    binding_requirements: list[dict[str, Any]] = Field(default_factory=list)


class ImportedWorkflowTemplate(BaseModel):
    """Recipient-owned disabled runtime workflow and required local bindings."""

    workflow: WorkflowDetail
    binding_requirements: list[dict[str, Any]]


class WorkflowTemplateProjectionService:
    """Owns one-way client template projection persistence and recipient import."""

    def __init__(self, workflow_service: WorkflowService) -> None:
        self.workflow_service = workflow_service
        self.repository = workflow_service.repository

    def upsert_projection(
        self,
        workflow_id: str,
        user_id: str,
        *,
        template_id: str,
        source_version: int,
        ciphertext: str,
        ciphertext_checksum: str,
        owner_wrapped_key: str,
        projection_schema_version: int,
    ) -> WorkflowTemplateProjectionRecord:
        """Persist client ciphertext without changing the runtime workflow state."""
        self.workflow_service.ensure_enabled()
        workflow = self.workflow_service.get_workflow(workflow_id, user_id)
        if source_version < workflow.version:
            raise WorkflowTemplateProjectionStaleError("source_version is older than the current workflow version")
        if source_version != workflow.version:
            raise WorkflowTemplateProjectionError("source_version must match the current workflow version")

        existing = self.repository.get_template_projection_for_workflow(workflow_id, user_id)
        if existing and existing["template_id"] != template_id:
            raise WorkflowTemplateProjectionError("A workflow projection must keep its stable template_id")

        now = int(time.time())
        record = WorkflowTemplateProjectionRecord(
            template_id=template_id,
            workflow_id=workflow_id,
            owner_hash=existing["owner_hash"] if existing else self.repository.workflow_owner_hash(user_id),
            source_version=source_version,
            projection_schema_version=projection_schema_version,
            ciphertext=ciphertext,
            ciphertext_checksum=ciphertext_checksum,
            owner_wrapped_key=owner_wrapped_key,
            created_at=existing["created_at"] if existing else now,
            updated_at=now,
            revoked_at=existing.get("revoked_at") if existing else None,
        )
        return WorkflowTemplateProjectionRecord.model_validate(self.repository.save_template_projection(record.model_dump()))

    def import_template(self, user_id: str, payload: dict[str, Any] | WorkflowTemplateImportPayload) -> ImportedWorkflowTemplate:
        """Create a fresh recipient-owned disabled runtime workflow from a portable template."""
        self.workflow_service.ensure_enabled()
        try:
            template = payload if isinstance(payload, WorkflowTemplateImportPayload) else WorkflowTemplateImportPayload.model_validate(payload)
        except ValidationError as exc:
            raise WorkflowTemplateImportError(str(exc)) from exc

        template_data = template.model_dump(mode="json")
        _reject_forbidden_template_fields(template_data)
        graph = _graph_from_template(template)
        _validate_declared_capabilities(graph, template.required_capabilities)
        _validate_binding_requirements(graph, template.binding_requirements)

        imported_graph, binding_requirements = _with_recipient_runtime_node_ids(graph)
        workflow = self.workflow_service.create_workflow(
            user_id,
            template.title,
            imported_graph,
            enabled=False,
            source="import",
            description=template.description,
        )
        return ImportedWorkflowTemplate(workflow=workflow, binding_requirements=binding_requirements)


def _reject_forbidden_template_fields(value: Any, path: str = "$") -> None:
    if isinstance(value, dict):
        for key, child in value.items():
            normalized_key = "".join(character for character in str(key).lower() if character.isalnum())
            if any(forbidden in normalized_key for forbidden in FORBIDDEN_TEMPLATE_FIELD_PARTS):
                raise WorkflowTemplateImportError(f"Workflow template contains forbidden field at {path}.{key}")
            _reject_forbidden_template_fields(child, f"{path}.{key}")
    elif isinstance(value, list):
        for index, child in enumerate(value):
            _reject_forbidden_template_fields(child, f"{path}[{index}]")


def _graph_from_template(template: WorkflowTemplateImportPayload) -> WorkflowGraph:
    try:
        return WorkflowGraph.model_validate(
            {
                "version": template.template_version,
                "trigger_node_id": template.trigger_template.get("id"),
                "nodes": [template.trigger_template, *template.node_templates],
                "edges": template.edge_templates,
                "variables": {"schema": template.variables_schema},
            }
        )
    except ValidationError as exc:
        raise WorkflowTemplateImportError(str(exc)) from exc


def _required_capabilities(graph: WorkflowGraph) -> set[str]:
    capabilities: set[str] = set()
    for node in graph.nodes:
        if node.type != WorkflowNodeType.APP_SKILL_ACTION:
            continue
        app_id = str(node.config.get("app_id") or "").strip()
        skill_id = str(node.config.get("skill_id") or "").strip()
        if app_id:
            capabilities.add(app_id)
        if app_id and skill_id:
            capabilities.add(f"{app_id}.{skill_id}")
    return capabilities


def _binding_requirements(graph: WorkflowGraph) -> list[dict[str, Any]]:
    requirements: list[dict[str, Any]] = []
    for node in graph.nodes:
        if node.type == WorkflowNodeType.SCHEDULE_TRIGGER:
            requirements.append({"type": "schedule", "node_id": node.id})
        elif node.type == WorkflowNodeType.APP_SKILL_ACTION:
            requirements.append(
                {
                    "type": "app_skill",
                    "node_id": node.id,
                    "app_id": node.config.get("app_id"),
                    "skill_id": node.config.get("skill_id"),
                }
            )
        elif node.type in {WorkflowNodeType.SEND_NOTIFICATION, WorkflowNodeType.SEND_EMAIL_NOTIFICATION}:
            requirements.append({"type": "notification_preferences", "node_id": node.id})
    return requirements


def _validate_declared_capabilities(graph: WorkflowGraph, declared_capabilities: list[str]) -> None:
    missing = _required_capabilities(graph).difference(declared_capabilities)
    if missing:
        raise WorkflowTemplateImportError(f"Workflow template is missing required capability declarations: {sorted(missing)}")


def _validate_binding_requirements(graph: WorkflowGraph, declared_requirements: list[dict[str, Any]]) -> None:
    expected = {(item["type"], item["node_id"]) for item in _binding_requirements(graph)}
    declared = {
        (str(item.get("type") or ""), str(item.get("node_id") or ""))
        for item in declared_requirements
        if isinstance(item, dict)
    }
    missing = expected.difference(declared)
    if missing:
        raise WorkflowTemplateImportError(f"Workflow template is missing binding requirements: {sorted(missing)}")


def _with_recipient_runtime_node_ids(graph: WorkflowGraph) -> tuple[WorkflowGraph, list[dict[str, Any]]]:
    node_id_map = {node.id: str(uuid.uuid4()) for node in graph.nodes}
    graph_data = graph.model_dump(mode="json", by_alias=True)
    graph_data["trigger_node_id"] = node_id_map[graph.trigger_node_id]
    for node in graph_data["nodes"]:
        node["id"] = node_id_map[node["id"]]
    for edge in graph_data["edges"]:
        edge["from"] = node_id_map[edge["from"]]
        edge["to"] = node_id_map[edge["to"]]
    imported_graph = WorkflowGraph.model_validate(graph_data)
    return imported_graph, _binding_requirements(imported_graph)

File: frontend/packages/ui/src/services/workflowTemplateService.ts
// frontend/packages/ui/src/services/workflowTemplateService.ts
// Client-side encryption and transport for portable Workflow templates.
// Runtime Workflows stay server-authoritative; this module only creates a
// recursively validated, portable projection for sharing and import.
// Fragment keys are generated and consumed locally and never sent to the API.

import { getApiEndpoint } from "../config/api";
import { decryptChatKeyWithMasterKey, encryptChatKeyWithMasterKey } from "./cryptoService";
import { buildShortUrl, encryptShareUrl, generateShortUrlParts } from "./shortUrlEncryption";
import { workflowTemplateProjectionStore, type WorkflowTemplateProjectionLocalRecord } from "../stores/workflowTemplateProjectionStore";
import type { WorkflowDetail, WorkflowGraph, WorkflowNode } from "../stores/workflowWorkspaceStore";

const PROJECTION_SCHEMA_VERSION = 1;
const TEMPLATE_KEY_LENGTH = 32;
const AES_GCM_IV_LENGTH = 12;
const FORBIDDEN_FIELD_PARTS = [
  "token", "secret", "credential", "accountid", "connectionid", "connectedaccountid",
  "provideruserid", "webhooksecret", "apikey", "password", "vault", "grant", "runid",
  "versionid", "workflowid", "nextrunat", "claim", "wait", "output", "providerresponse",
  "sourcechatid", "encryptedgraphblobref", "encryptedcontentref", "fragmentkey", "shortkey",
  "templatekey",
];

export type WorkflowTemplateBindingRequirement = {
  type: string;
  node_id: string;
  app_id?: string;
  skill_id?: string;
};

export type WorkflowTemplatePayload = {
  template_version: number;
  title: string;
  description?: string | null;
  trigger_template: Record<string, unknown>;
  node_templates: Array<Record<string, unknown>>;
  edge_templates: Array<Record<string, unknown>>;
  variables_schema: Record<string, unknown>;
  required_capabilities: string[];
  binding_requirements: WorkflowTemplateBindingRequirement[];
};

export type ImportedWorkflowTemplate = WorkflowDetail & {
  binding_requirements: WorkflowTemplateBindingRequirement[];
};

type PublicProjectionResponse = {
  template_id: string;
  ciphertext: string;
  ciphertext_checksum: string;
  projection_schema_version: number;
};

type ShortShareResult = {
  shortUrl: string;
  longUrl: string;
  token: string;
};

function encodeBase64Url(bytes: Uint8Array): string {
  let binary = "";
  for (const byte of Array.from(bytes)) binary += String.fromCharCode(byte);
  return btoa(binary).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

function decodeBase64Url(value: string): Uint8Array {
  if (!/^[A-Za-z0-9_-]+$/.test(value)) throw new Error("Template key or ciphertext is malformed.");
  let base64 = value.replace(/-/g, "+").replace(/_/g, "/");
  while (base64.length % 4) base64 += "=";
  const binary = atob(base64);
  return Uint8Array.from(binary, (character) => character.charCodeAt(0));
}

function cloneJson<T>(value: T): T {
  return JSON.parse(JSON.stringify(value)) as T;
}

function normalizedKey(key: string): string {
  return key.toLowerCase().replace(/[^a-z0-9]/g, "");
}

function assertPortableValue(value: unknown, path = "$"): void {
  if (Array.isArray(value)) {
    value.forEach((item, index) => assertPortableValue(item, `${path}[${index}]`));
    return;
  }
  if (typeof value !== "object" || value === null) return;
  for (const [key, child] of Object.entries(value)) {
    const normalized = normalizedKey(key);
    if (FORBIDDEN_FIELD_PARTS.some((forbidden) => normalized.includes(forbidden))) {
      throw new Error(`Workflow template contains a non-portable field at ${path}.${key}.`);
    }
    assertPortableValue(child, `${path}.${key}`);
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function isPortableNode(value: unknown): value is Record<string, unknown> {
  return isRecord(value)
    && typeof value.id === "string"
    && value.id.length > 0
    && typeof value.type === "string"
    && value.type.length > 0
    && isRecord(value.config);
}

function isPortableEdge(value: unknown): value is Record<string, unknown> {
  return isRecord(value)
    && typeof value.from === "string"
    && value.from.length > 0
    && typeof value.to === "string"
    && value.to.length > 0
    && (value.branch === undefined || typeof value.branch === "string");
}

function isBindingRequirement(value: unknown): value is WorkflowTemplateBindingRequirement {
  return isRecord(value)
    && typeof value.type === "string"
    && value.type.length > 0
    && typeof value.node_id === "string"
    && value.node_id.length > 0
    && (value.app_id === undefined || typeof value.app_id === "string")
    && (value.skill_id === undefined || typeof value.skill_id === "string");
}

function portableNode(node: WorkflowNode): Record<string, unknown> {
  const candidate = {
    id: node.id,
    type: node.type,
    ...(node.title ? { title: node.title } : {}),
    config: cloneJson(node.config ?? {}),
  };
  assertPortableValue(candidate, `$.nodes.${node.id}`);
  return candidate;
}

function bindingRequirements(graph: WorkflowGraph): WorkflowTemplateBindingRequirement[] {
  return graph.nodes.flatMap((node) => {
    if (node.type === "schedule_trigger") return [{ type: "schedule", node_id: node.id }];
    if (node.type === "app_skill_action") {
      const config = node.config ?? {};
      return [{
        type: "app_skill",
        node_id: node.id,
        app_id: String(config.app_id ?? ""),
        skill_id: String(config.skill_id ?? ""),
      }];
    }
    if (node.type === "send_notification" || node.type === "send_email_notification") {
      return [{ type: "notification_preferences", node_id: node.id }];
    }
    return [];
  });
}

function requiredCapabilities(requirements: WorkflowTemplateBindingRequirement[]): string[] {
  const capabilities = new Set<string>();
  for (const requirement of requirements) {
    if (requirement.type !== "app_skill" || !requirement.app_id) continue;
    capabilities.add(requirement.app_id);
    if (requirement.skill_id) capabilities.add(`${requirement.app_id}.${requirement.skill_id}`);
  }
  return Array.from(capabilities).sort();
}

function triggerNode(graph: WorkflowGraph): WorkflowNode {
  const trigger = graph.nodes.find((node) => node.id === graph.trigger_node_id);
  if (!trigger) throw new Error("Workflow does not have a portable trigger node.");
  return trigger;
}

export function buildWorkflowTemplatePayload(workflow: WorkflowDetail): WorkflowTemplatePayload {
  const trigger = triggerNode(workflow.graph);
  const requirements = bindingRequirements(workflow.graph);
  const payload: WorkflowTemplatePayload = {
    template_version: workflow.graph.version,
    title: workflow.title.trim(),
    description: workflow.description?.trim() || null,
    trigger_template: portableNode(trigger),
    node_templates: workflow.graph.nodes.filter((node) => node.id !== trigger.id).map(portableNode),
    edge_templates: workflow.graph.edges.map((edge) => cloneJson({ from: edge.from, to: edge.to, ...(edge.branch ? { branch: edge.branch } : {}) })),
    variables_schema: {},
    required_capabilities: requiredCapabilities(requirements),
    binding_requirements: requirements,
  };
  if (!payload.title) throw new Error("Workflow title is required before sharing.");
  assertPortableValue(payload);
  return payload;
}

export function validateWorkflowTemplatePayload(value: unknown): WorkflowTemplatePayload {
  if (!isRecord(value) || typeof value.template_version !== "number" || !Number.isInteger(value.template_version) || value.template_version < 1) {
    throw new Error("Workflow template has an invalid version.");
  }
  if (typeof value.title !== "string" || !value.title.trim() || !isPortableNode(value.trigger_template)) {
    throw new Error("Workflow template is missing required portable fields.");
  }
  const nodeTemplates = Array.isArray(value.node_templates) ? value.node_templates : null;
  const edgeTemplates = Array.isArray(value.edge_templates) ? value.edge_templates : null;
  const variablesSchema = isRecord(value.variables_schema) ? value.variables_schema : null;
  const capabilities = Array.isArray(value.required_capabilities) && value.required_capabilities.every((item) => typeof item === "string")
    ? value.required_capabilities
    : null;
  const requirements = Array.isArray(value.binding_requirements) && value.binding_requirements.every(isBindingRequirement)
    ? value.binding_requirements
    : null;
  if (!nodeTemplates || !nodeTemplates.every(isPortableNode) || !edgeTemplates || !edgeTemplates.every(isPortableEdge) || !variablesSchema || !capabilities || !requirements) {
    throw new Error("Workflow template has an invalid portable structure.");
  }
  assertPortableValue(value);
  return value as WorkflowTemplatePayload;
}

async function checksum(value: string): Promise<string> {
  const digest = await crypto.subtle.digest("SHA-256", new TextEncoder().encode(value));
  return `sha256:${encodeBase64Url(new Uint8Array(digest))}`;
}

function generateTemplateId(): string {
  return `wt_${crypto.randomUUID().replace(/-/g, "")}`;
}

function generateTemplateKey(): Uint8Array {
  return crypto.getRandomValues(new Uint8Array(TEMPLATE_KEY_LENGTH));
}

async function encryptProjection(payload: WorkflowTemplatePayload, keyBytes: Uint8Array): Promise<string> {
  const key = await crypto.subtle.importKey("raw", new Uint8Array(keyBytes), { name: "AES-GCM" }, false, ["encrypt"]);
  const iv = crypto.getRandomValues(new Uint8Array(AES_GCM_IV_LENGTH));
  const encrypted = await crypto.subtle.encrypt({ name: "AES-GCM", iv }, key, new TextEncoder().encode(JSON.stringify(payload)));
  const combined = new Uint8Array(iv.length + encrypted.byteLength);
  combined.set(iv);
  combined.set(new Uint8Array(encrypted), iv.length);
  return encodeBase64Url(combined);
}

async function decryptProjection(ciphertext: string, encodedKey: string): Promise<WorkflowTemplatePayload> {
  const keyBytes = decodeBase64Url(encodedKey);
  if (keyBytes.length !== TEMPLATE_KEY_LENGTH) throw new Error("Workflow template key has an invalid length.");
  const combined = decodeBase64Url(ciphertext);
  if (combined.length <= AES_GCM_IV_LENGTH) throw new Error("Workflow template ciphertext is incomplete.");
  const key = await crypto.subtle.importKey("raw", new Uint8Array(keyBytes), { name: "AES-GCM" }, false, ["decrypt"]);
  const decrypted = await crypto.subtle.decrypt({ name: "AES-GCM", iv: combined.slice(0, AES_GCM_IV_LENGTH) }, key, combined.slice(AES_GCM_IV_LENGTH));
  return validateWorkflowTemplatePayload(JSON.parse(new TextDecoder().decode(decrypted)) as unknown);
}

async function workflowRequest<T>(path: string, init: RequestInit = {}): Promise<T> {
  const headers = new Headers(init.headers);
  headers.set("Accept", "application/json");
  headers.set("Content-Type", "application/json");
  const response = await fetch(getApiEndpoint(path), {
    ...init,
    credentials: "include",
    headers,
  });
  if (!response.ok) throw new Error(`Workflow template request failed with HTTP ${response.status}.`);
  return await response.json() as T;
}

async function localTemplateKey(record: WorkflowTemplateProjectionLocalRecord): Promise<string> {
  const key = await decryptChatKeyWithMasterKey(record.ownerWrappedKey);
  if (!key || key.length !== TEMPLATE_KEY_LENGTH) throw new Error("Template key is unavailable on this device. Create a new share from the owner device.");
  return encodeBase64Url(key);
}

export async function upsertWorkflowTemplateProjection(workflow: WorkflowDetail): Promise<WorkflowTemplateProjectionLocalRecord> {
  const existing = workflowTemplateProjectionStore.get(workflow.id);
  const templateId = existing?.templateId ?? generateTemplateId();
  const keyBytes = existing ? decodeBase64Url(await localTemplateKey(existing)) : generateTemplateKey();
  const ownerWrappedKey = existing?.ownerWrappedKey ?? await encryptChatKeyWithMasterKey(keyBytes);
  if (!ownerWrappedKey) throw new Error("Your encryption key is unavailable. Re-authenticate before sharing this workflow.");
  const ciphertext = await encryptProjection(buildWorkflowTemplatePayload(workflow), keyBytes);
  const ciphertextChecksum = await checksum(ciphertext);
  const result = await workflowRequest<{ template_id: string; source_version: number; updated_at: number }>(
    `/v1/workflows/${encodeURIComponent(workflow.id)}/template-projection`,
    {
      method: "PUT",
      body: JSON.stringify({
        template_id: templateId,
        source_version: workflow.version ?? 1,
        ciphertext,
        ciphertext_checksum: ciphertextChecksum,
        owner_wrapped_key: ownerWrappedKey,
        projection_schema_version: PROJECTION_SCHEMA_VERSION,
      }),
    },
  );
  const record: WorkflowTemplateProjectionLocalRecord = {
    workflowId: workflow.id,
    templateId: result.template_id,
    sourceVersion: result.source_version,
    ciphertext,
    ciphertextChecksum,
    ownerWrappedKey,
    projectionSchemaVersion: PROJECTION_SCHEMA_VERSION,
    shortToken: existing?.shortToken,
    revokedAt: existing?.revokedAt ?? null,
    updatedAt: result.updated_at,
  };
  workflowTemplateProjectionStore.upsert(record);
  return record;
}

export async function createWorkflowTemplateShare(workflow: WorkflowDetail): Promise<ShortShareResult> {
  const record = await upsertWorkflowTemplateProjection(workflow);
  const longUrl = await buildWorkflowTemplateLongUrl(record);
  const { token, shortKey } = generateShortUrlParts();
  const encryptedUrl = await encryptShareUrl(longUrl, token, shortKey);
  await workflowRequest("/v1/share/short-url", {
    method: "POST",
    body: JSON.stringify({
      token,
      encrypted_url: encryptedUrl,
      content_type: "workflow_template",
      content_id: record.templateId,
      password_protected: false,
    }),
  });
  workflowTemplateProjectionStore.upsert({ ...record, shortToken: token, revokedAt: null, updatedAt: Date.now() });
  return { shortUrl: buildShortUrl(token, shortKey), longUrl, token };
}

export async function buildWorkflowTemplateLongUrl(record: WorkflowTemplateProjectionLocalRecord): Promise<string> {
  const templateKey = await localTemplateKey(record);
  const origin = typeof window === "undefined" ? "https://openmates.org" : window.location.origin;
  return `${origin}/share/workflow-template/${encodeURIComponent(record.templateId)}#key=${templateKey}`;
}

export async function revokeWorkflowTemplateShare(workflowId: string): Promise<void> {
  const record = workflowTemplateProjectionStore.get(workflowId);
  await workflowRequest(`/v1/workflows/${encodeURIComponent(workflowId)}/template-projection/revoke`, { method: "POST", body: "{}" });
  if (record?.shortToken) {
    await workflowRequest(`/v1/share/short-url/${encodeURIComponent(record.shortToken)}`, { method: "DELETE" });
  }
  workflowTemplateProjectionStore.setRevoked(workflowId, Date.now());
}

export async function unrevokeWorkflowTemplateShare(workflowId: string): Promise<void> {
  await workflowRequest(`/v1/workflows/${encodeURIComponent(workflowId)}/template-projection/unrevoke`, { method: "POST", body: "{}" });
  workflowTemplateProjectionStore.setRevoked(workflowId, null);
}

export async function loadSharedWorkflowTemplate(templateId: string, templateKey: string): Promise<WorkflowTemplatePayload> {
  const projection = await workflowRequest<PublicProjectionResponse>(`/v1/workflows/template-projections/${encodeURIComponent(templateId)}`);
  if (projection.template_id !== templateId || projection.projection_schema_version !== PROJECTION_SCHEMA_VERSION) {
    throw new Error("Workflow template projection is not compatible with this app version.");
  }
  if (projection.ciphertext_checksum !== await checksum(projection.ciphertext)) {
    throw new Error("Workflow template integrity check failed.");
  }
  return await decryptProjection(projection.ciphertext, templateKey);
}

export async function importWorkflowTemplate(payload: WorkflowTemplatePayload): Promise<ImportedWorkflowTemplate> {
  const response = await workflowRequest<{ workflow: ImportedWorkflowTemplate }>("/v1/workflows/template-import", {
    method: "POST",
    body: JSON.stringify(validateWorkflowTemplatePayload(payload)),
  });
  if (!response.workflow || response.workflow.enabled) throw new Error("Imported workflow did not remain disabled for rebinding.");
  return response.workflow;
}

export async function completeWorkflowTemplateBinding(workflowId: string, requirement: WorkflowTemplateBindingRequirement): Promise<void> {
  await workflowRequest(`/v1/workflows/${encodeURIComponent(workflowId)}/binding-requirements/complete`, {
    method: "POST",
    body: JSON.stringify({ type: requirement.type, node_id: requirement.node_id }),
  });
}
